presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayContains.java

```diff
@@ -13,120 +13,46 @@
-    {
-        return true;
-    }
-
-    @Override
-    public String getDescription()
-    {
-        return "Determines whether given value exists in the array";
-    }
+    private ArrayContains() {}
 
-    @Override
-    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
-    {
-        Type type = boundVariables.getTypeVariable("T");
-
-        MethodHandle methodHandle;
-        MethodHandle equalsHandle = functionRegistry.getScalarFunctionImplementation(internalOperator(OperatorType.EQUAL, BooleanType.BOOLEAN, ImmutableList.of(type, type))).getMethodHandle();
-
-        List<Boolean> nullableArguments;
-        if (type.getJavaType() == void.class) {
-            nullableArguments = ImmutableList.of(false, true);
-            methodHandle = METHOD_HANDLE_UNKNOWN;
-        }
-        else {
-            nullableArguments = ImmutableList.of(false, false);
-            methodHandle = methodHandle(ArrayContains.class, "contains", Type.class, MethodHandle.class, Block.class, type.getJavaType());
-        }
-
-        return new ScalarFunctionImplementation(true, nullableArguments, methodHandle.bindTo(type).bindTo(equalsHandle), isDeterministic());
-    }
-
-    public static Boolean arrayWithUnknownType(Type elementType, MethodHandle equals, Block arrayBlock, Void value)
-    {
-        return null;
-    }
-
-    public static Boolean contains(Type elementType, MethodHandle equals, Block arrayBlock, Block value)
+    private static <R> Boolean containsImplementation(MethodHandle equalsFunction, IntFunction<R> reader, Block array, R value)
     {
         boolean foundNull = false;
-        for (int i = 0; i < arrayBlock.getPositionCount(); i++) {
-            if (arrayBlock.isNull(i)) {
-                foundNull = true;
-                continue;
-            }
-            try {
-                if ((boolean) equals.invokeExact((Block) elementType.getObject(arrayBlock, i), value)) {
-                    return true;
-                }
-            }
-            catch (Throwable t) {
-                Throwables.propagateIfInstanceOf(t, Error.class);
-                Throwables.propagateIfInstanceOf(t, PrestoException.class);
-
-                throw new PrestoException(INTERNAL_ERROR, t);
-            }
-        }
-        if (foundNull) {
+        int elementCount = array.getPositionCount();
+        if (value == null) {
             return null;
         }
-        return false;
-    }
-
-    public static Boolean contains(Type elementType, MethodHandle equals, Block arrayBlock, Slice value)
-    {
-        boolean foundNull = false;
-        for (int i = 0; i < arrayBlock.getPositionCount(); i++) {
-            if (arrayBlock.isNull(i)) {
+        for (int i = 0; i < elementCount; i++) {
+            if (array.isNull(i)) {
                 foundNull = true;
                 continue;
             }
+
+            R element = reader.apply(i);
```

Comment:
I think this may have performance problems due to type erasure and no longer using `invokeExact` below. Also, this should be split into two commits. One that migrates to the new framework, and a second that refactors the implementation. For the refactoring, can you add a benchmark like `BenchmarkArrayHashCodeOperator`?